### PR TITLE
client: only wait for write MDS OPs when unmounting

### PR DESCRIPTION
**Pull request by lxbsz**
We do not care about the read MDS OPs and it's safe by just dropping
them when unmounting.

Fixes: https://tracker.ceph.com/issues/59705
Signed-off-by: Xiubo Li <xiubli@redhat.com>

Sync with the kclient.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -6650,10 +6650,18 @@ void Client::_unmount(bool abort)
 
   mount_cond.wait(lock, [this] {
     if (!mds_requests.empty()) {
```
Comment (dparmar18) [inline src/client/Client.cc:6652]: If we return from here if the `mds_requests` is non-empty, then the below code would not be executed, also the removal of line `return mds_requests.empty();` indicates this line should be `if(mds_requests.empty())`?

File: src/client/Client.cc [hunk 2]
```diff
@@ -6650,10 +6650,18 @@ void Client::_unmount(bool abort)
...
-		     << dendl;
+      return true;
     }
-    return mds_requests.empty();
+
+    // Only wait for write OPs
+    for (auto& [tid, req] : mds_requests) {
+      if (req->is_write()) {
+        ldout(cct, 10) << "waiting on " << mds_requests.size() << " requests"
+                       << dendl;
```
Comment (dparmar18) [inline src/client/Client.cc:6660]: this could be
```
ldout(cct, 10) << tid << " waiting for write requests to complete"
                        << dendl;
```

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc]: Yeah, you are right. It should be `if(mds_requests.empty())`, I will fix it. thanks @dparmar18

File: src/client/Client.cc (hunk 2 above)
Comment (lxbsz) [reply src/client/Client.cc]: Looks better.

File: src/client/Client.cc [hunk 3]
```diff
@@ -6649,11 +6649,16 @@ void Client::_unmount(bool abort)
   }
 
   mount_cond.wait(lock, [this] {
-    if (!mds_requests.empty()) {
-      ldout(cct, 10) << "waiting on " << mds_requests.size() << " requests"
-		     << dendl;
+    // Only wait for write OPs
+    for (auto& [tid, req] : mds_requests) {
+      if (req->is_write()) {
+        ldout(cct, 10) << "waiting for write requests to complete, "
```
Comment (neesingh-rh) [inline src/client/Client.cc:6655]: Logging here along with the request tid would be better.

Comment (lxbsz) [reply src/client/Client.cc:6655]: Make sense to me. Thanks @neesingh-rh

File: src/client/Client.cc [hunk 4]
```diff
@@ -6649,11 +6649,17 @@ void Client::_unmount(bool abort)
...
-    if (!mds_requests.empty()) {
-      ldout(cct, 10) << "waiting on " << mds_requests.size() << " requests"
-		     << dendl;
+    // Only wait for write OPs
+    for (auto& [tid, req] : mds_requests) {
+      if (req->is_write()) {
+        ldout(cct, 10) << "waiting for write request '" << tid
+                       << "' to complete, currently there are still "
+                       << mds_requests.size() << " read/write requests"
+                       << dendl;
```
Comment (dparmar18) [inline src/client/Client.cc:6658]: the line after comma looks a bit irrelevant, if that line is important piece of data then might be good to add it as 
`ldout(cct, 10) << "Current outstanding requests: " << mds_requests.size;()` before the `for` loop starts. something like
```
if(!mds_requests.empty()) {
    ldout(cct, 10) << "Current outstanding requests: " << mds_requests.size;   
}
```
and leave the line to just `ldout(cct, 10) << tid << " waiting for write request to complete";`

what do you think?

Comment (lxbsz) [reply src/client/Client.cc:6658]: > the line after comma looks a bit irrelevant, if that line is important piece of data then might be good to add it as `ldout(cct, 10) << "Current outstanding requests: " << mds_requests.size;()` before the `for` loop starts. something like
> 
> ```
> if(!mds_requests.empty()) {
>     ldout(cct, 10) << "Current outstanding requests: " << mds_requests.size;   
> }
> ```

This will print a log if the `mds_requests` is non-empty but they are all `read` requests, which seems unnecessary here.

> 
> and leave the line to just `ldout(cct, 10) << tid << " waiting for write request to complete";`
> 

This seems confusing for me. `waiting for write request '${tid}' to complete` looks much better.

> what do you think?

Comment (dparmar18) [reply src/client/Client.cc]: > This will print a log if the mds_requests is non-empty but they are all read requests, which seems unnecessary here.

makes sense

> This seems confusing for me. waiting for write request '${tid}' to complete looks much better.

Yeah this is better

**Review by batrick — APPROVED** [review]

File: src/client/Client.cc [hunk 5]
```diff
@@ -6649,11 +6649,18 @@ void Client::_unmount(bool abort)
...
 
   mount_cond.wait(lock, [this] {
-    if (!mds_requests.empty()) {
-      ldout(cct, 10) << "waiting on " << mds_requests.size() << " requests"
-		     << dendl;
+    // Only wait for write OPs
+    for (auto& [tid, req] : mds_requests) {
+      if (req->is_write()) {
+        ldout(cct, 10) << "waiting for write request '" << tid
+                       << "' to complete, current there are "
```
Comment (vshankar) [inline src/client/Client.cc:6656]: nit: currently there are....

**Review by vshankar — APPROVED** [review]
Minor nit, otherwise LGTM.

**Comment by vshankar** [comment]
@lxbsz one of the commits has missing signed-off-by.

File: src/client/Client.cc (hunk 5 above)
Comment (lxbsz) [reply src/client/Client.cc]: Fixed it. Thanks.

**Comment by lxbsz** [comment]
> @lxbsz one of the commits has missing signed-off-by.

Both had added this and I have no idea why that failed before.

**Review by dparmar18 — APPROVED** [review]

**Comment by dparmar18** [comment]
> > @lxbsz one of the commits has missing signed-off-by.
> 
> Both had added this and I have no idea why that failed before.

I've seen this in my PR too, something's wrong with jenkins

**Comment by lxbsz** [comment]
> > > @lxbsz one of the commits has missing signed-off-by.
> > 
> > 
> > Both had added this and I have no idea why that failed before.
> 
> I've seen this in my PR too, something's wrong with jenkins

yeah, it seems.

**Comment by vshankar** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test make check

**Comment by dparmar18** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#22-June-2023
